File: ag-ui/integrations/langgraph/typescript/src/interrupts.ts
import type { Interrupt as LangGraphInterrupt } from "@langchain/langgraph-sdk";
import type {
  Interrupt as AGUIInterrupt,
  ResumeEntry,
  RunAgentInput,
} from "@ag-ui/core";

const isPlainObject = (v: unknown): v is Record<string, unknown> =>
  typeof v === "object" && v !== null && !Array.isArray(v);

export function langGraphInterruptToAGUI(
  lg: LangGraphInterrupt,
): AGUIInterrupt {
  const raw = lg.value;
  const dict = isPlainObject(raw) ? raw : null;

  if (!lg.id) {
    throw new Error(
      "LangGraph Interrupt is missing `id`. The id is required to match a " +
        "resume answer back to the originating step; synthesising an id here " +
        "would silently misroute multi-interrupt resumes. Upgrade to " +
        "@langchain/langgraph-sdk that always populates Interrupt.id.",
    );
  }
  const id = lg.id;
  const reason =
    (dict?.reason as string | undefined) ?? "langgraph:interrupt";

  const message =
    typeof raw === "string"
      ? raw
      : (dict?.message as string | undefined);
  const toolCallId =
    (dict?.toolCallId as string | undefined) ??
    (dict?.tool_call_id as string | undefined);
  const responseSchema =
    (dict?.responseSchema as Record<string, unknown> | undefined) ??
    (dict?.response_schema as Record<string, unknown> | undefined);
  const expiresAt =
    (dict?.expiresAt as string | undefined) ??
    (dict?.expires_at as string | undefined);

  const metadata: Record<string, unknown> = {
    langgraph: {
      raw,
      ns: (lg as { ns?: string[] }).ns,
      resumable: (lg as { resumable?: boolean }).resumable,
      when: (lg as { when?: string }).when,
    },
  };

  return {
    id,
    reason,
    ...(message !== undefined ? { message } : {}),
    ...(toolCallId !== undefined ? { toolCallId } : {}),
    ...(responseSchema !== undefined ? { responseSchema } : {}),
    ...(expiresAt !== undefined ? { expiresAt } : {}),
    metadata,
  };
}

export function langGraphInterruptsToAGUI(
  list: readonly LangGraphInterrupt[],
): AGUIInterrupt[] {
  return list.map(langGraphInterruptToAGUI);
}

/**
 * Detect whether a run is a *legacy* resume — i.e. the caller is resuming a
 * LangGraph interrupt via the deprecated `forwardedProps.command.resume`
 * channel and has NOT populated the canonical `RunAgentInput.resume[]`.
 *
 * Released clients (e.g. CopilotKit's `useLangGraphInterrupt`) still resume
 * this way. When the integration emits `RUN_FINISHED.outcome=interrupt`
 * (`emitInterruptOutcome` enabled, or `enableLegacyOnInterruptEvent` off — which
 * forces the outcome), `AbstractAgent` records `pendingInterrupts`, and the base
 * `onInitialize` lifecycle would otherwise reject the legacy resume run with
 * "pending interrupt(s) not addressed by resume" — a regression for every such
 * legacy client. See `reconcileLegacyResumeInterrupts`. (With the default
 * config no outcome is emitted, `pendingInterrupts` stays empty, and the bridge
 * is a no-op.)
 */
export function isLegacyCommandResume(input: RunAgentInput): boolean {
  const legacyResume = (input.forwardedProps as Record<string, any> | undefined)
    ?.command?.resume;
  const hasAguiResume =
    Array.isArray(input.resume) && input.resume.length > 0;
  return legacyResume !== undefined && !hasAguiResume;
}

/**
 * Back-compat bridge invoked from `onInitialize` before the base lifecycle
 * runs. When a run is a legacy `command.resume` resume (see
 * `isLegacyCommandResume`), drop the agent's tracked `pendingInterrupts` so the
 * base "uncovered interrupt" guard does not reject it. The legacy resume is
 * carried through `forwardedProps.command.resume` and resolved by
 * `runAgentStream` exactly as it was before structured interrupts existed, so
 * the tracked list is not needed for this run.
 *
 * Note: this intentionally also bypasses the base lifecycle's interrupt-expiry
 * check for legacy resumes — the legacy `command.resume` channel never enforced
 * expiry (the graph resolves the resume itself), so this preserves pre-existing
 * behavior rather than introducing a new gap.
 */
export function reconcileLegacyResumeInterrupts(
  agent: { pendingInterrupts: AGUIInterrupt[] },
  input: RunAgentInput,
): void {
  if (agent.pendingInterrupts.length > 0 && isLegacyCommandResume(input)) {
    agent.pendingInterrupts = [];
  }
}

export const DEFAULT_RESUME_SENTINEL_CANCELLED = "__agui_cancelled__";
export const DEFAULT_RESUME_SENTINEL_MAP = "__agui_resume_map__";

export function buildLgCommandResumeFromAgui(
  entries: readonly ResumeEntry[],
): unknown {
  if (entries.length === 1) {
    const e = entries[0];
    if (e.status === "resolved") return e.payload;
    return { [DEFAULT_RESUME_SENTINEL_CANCELLED]: true, interrupt_id: e.interruptId };
  }
  return {
    [DEFAULT_RESUME_SENTINEL_MAP]: Object.fromEntries(
      entries.map((e) => [
        e.interruptId,
        { status: e.status, payload: e.payload ?? null },
      ]),
    ),
  };
}

File: ag-ui/integrations/langgraph/python/ag_ui_langgraph/interrupts.py
from typing import Any, List, Mapping

from ag_ui.core import Interrupt as AGUIInterrupt
from langgraph.types import Interrupt as LangGraphInterrupt

from .utils import make_json_safe


def _first_not_none(*values):
    return next((v for v in values if v is not None), None)


def lg_interrupt_to_agui(lg: LangGraphInterrupt) -> AGUIInterrupt:
    raw = lg.value
    is_dict = isinstance(raw, Mapping)

    interrupt_id = getattr(lg, "id", None)
    if not interrupt_id:
        raise ValueError(
            "LangGraph Interrupt is missing `id`. The id is required to "
            "match a resume answer back to the originating step; synthesising "
            "an id here would silently misroute multi-interrupt resumes. "
            "Upgrade to langgraph>=0.4 (which always populates Interrupt.id)."
        )

    # Default only when reason is absent (None), not when it is a falsy-but-real
    # value: an explicit reason="" must be preserved, matching the TS side's
    # `?? "langgraph:interrupt"`. Using `or` here would drop "".
    _reason = raw.get("reason") if is_dict else None
    reason = _reason if _reason is not None else "langgraph:interrupt"

    message = (
        raw if isinstance(raw, str)
        else raw.get("message") if is_dict else None
    )
    tool_call_id = _first_not_none(
        raw.get("toolCallId") if is_dict else None,
        raw.get("tool_call_id") if is_dict else None,
    )
    response_schema = _first_not_none(
        raw.get("responseSchema") if is_dict else None,
        raw.get("response_schema") if is_dict else None,
    )
    expires_at = _first_not_none(
        raw.get("expiresAt") if is_dict else None,
        raw.get("expires_at") if is_dict else None,
    )

    metadata: dict[str, Any] = {
        "langgraph": {
            "raw": make_json_safe(raw),
            "ns": getattr(lg, "ns", None),
            "resumable": getattr(lg, "resumable", None),
            "when": getattr(lg, "when", None),
        }
    }

    return AGUIInterrupt(
        id=interrupt_id,
        reason=reason,
        message=message,
        tool_call_id=tool_call_id,
        response_schema=response_schema,
        expires_at=expires_at,
        metadata=metadata,
    )


def lg_interrupts_to_agui(items) -> List[AGUIInterrupt]:
    return [lg_interrupt_to_agui(i) for i in items]


DEFAULT_RESUME_SENTINEL_CANCELLED = "__agui_cancelled__"
DEFAULT_RESUME_SENTINEL_MAP = "__agui_resume_map__"
